### PR TITLE
Hypothesis tests for roundtrip to & from pandas

### PR DESCRIPTION
**Pull request by takluyver**
Part of #1846: test roundtripping between xarray DataArray & Dataset and pandas Series & DataFrame.

I haven't particularly tried to hunt down corner cases (e.g. dataframes with 0 columns), in favour of adding tests that currently pass. But these tests probably form a useful platform if you do want to ensure corner cases like that behave nicely - just modify the limits and see what fails.

### CODE REVIEW

**Comment by max-sixty** [comment]
This looks great! I'll let someone who knows hypothesis better do a full review. Thanks for submitting @takluyver !

File: properties/conftest.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+from hypothesis import settings
+
+# Run for a while - arrays are a bigger search space than usual
+settings.register_profile("ci", deadline=None)
```
Comment (Zac-HD) [inline properties/conftest.py]: :+1: on moving the settings profile to `conftest.py`, I was about to suggest that on the NetCDF PR.

I'd add `print_blob=True` to this, which prints more details to allow local reproduction of failures in CI.  It's a bit more verbose, so we leave it disabled by default, but if you ever need it you'll be very glad it exists.

File: properties/test_pandas_roundtrip.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@given(numeric_homogeneous_dataframe)
+def test_roundtrip_pandas_dataframe(df):
+    # Need to name the indexes, otherwise Xarray names them 'dim_0', 'dim_1'.
+    df.index.name = "rows"
+    df.columns.name = "cols"
+    arr = xr.DataArray(df)
+    roundtripped = arr.to_pandas()
+    pd.testing.assert_frame_equal(df, roundtripped)
```
Comment (Zac-HD) [inline properties/test_pandas_roundtrip.py]: I would add another half-circuit to each of these tests, in this case `xr.testing.assert_identical(arr, roundtripped.to_xarray())`.

It seems redundant, but the pattern ocasionally picks up on odd cases where e.g. `.to_xarray()` can output a variety of objects which are all converted back to the same Pandas object.  I don't actually expect this to fail if the existing tests are passing, but it could in future.

File: properties/test_pandas_roundtrip.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+def datasets_1d_vars(draw):
+    """Generate datasets with only 1D variables
+
+    Suitable for converting to pandas dataframes.
+    """
+    n_vars = draw(st.integers(min_value=1, max_value=3))
+    n_entries = draw(st.integers(min_value=0, max_value=100))
+    dims = ("rows",)
+    vars = {}
+    for _ in range(n_vars):
```
Comment (Zac-HD) [inline properties/test_pandas_roundtrip.py]: This pattern - draw a number, then draw that many elements - is tempting but [tends to be inefficient](https://github.com/HypothesisWorks/hypothesis/blob/master/guides/strategies-that-shrink.rst) when Hypothesis tries to minimse any failures.

The alternative, which we recommend, is to generate collections using the `st.lists()` strategy - that way Hypothesis will be able to operate in terms of elements of the list.

In this case it's probably only worth doing so for either the vars or entries dimension, and keep the other as-is.  If you're keen to do both, it's complicated enough that I'd just fall back on the Hypothesis pandas extension and `.map(pd.DataFrame.to_xarray)` over the result :sweat_smile:

**Comment by takluyver** [comment]
As in my other PR, one suggested addition causes a test failure, and I've put that in the last commit.

File: properties/test_pandas_roundtrip.py (hunk 3 above)
Comment (Zac-HD) [reply properties/test_pandas_roundtrip.py]: You could draw `indices` before the loop, then draw a list of (name, array) tuples.  Or in this case you could use `st.dictionaries()` to, well, generate a list of key-value tuples internally.

The other nice trick would be to draw your index first, and use it's length - deleting elements from that will be *slightly* more efficient than shrinking the `n_elements` parameter.

Putting it all togther, I'd write 

```python
idx = draw(pdst.indexes(dtype="u8", min_size=0, max_size=100))
vars_strat = st.dictionaries(
    keys=st.text(),
    values=npst.arrays(dtype=numeric_dtypes, shape=len(idx)).map(partial(xr.Variable, ("rows",))),
    min_size=1,
    max_size=3,
)
return xr.Dataset(draw(vars_strat), coords={"rows": idx})
```

Comment (takluyver) [reply properties/test_pandas_roundtrip.py]: Thanks, that does look neater!

**Comment by max-sixty** [comment]
This seems so close—could we fix the test (maybe that's just a merging of master?) and merge?

**Comment by max-sixty** [comment]
This seems so close—could we fix the test (maybe that's just a merging of master?) and merge?

**Comment by takluyver** [comment]
Merged master, crossing fingers that fixes it.

**Comment by takluyver** [comment]
Nope. I don't understand the error, though it looks like astropy has had something similar: https://github.com/astropy/astropy/issues/6424

Also black is now failing on a number of files not affected here.

File: properties/test_encode_decode.py [hunk 4]
```diff
@@ -10,15 +10,10 @@
 
 import hypothesis.extra.numpy as npst
```
Comment (dcherian) [inline properties/test_encode_decode.py]: These may need to be guarded too using `pytest.importorskip` perhaps? @max-sixty  what do you think?

Comment (max-sixty) [reply properties/test_encode_decode.py]: Yes same as here! https://github.com/max-sixty/xarray/blob/black/properties/test_encode_decode.py#L9

Comment (takluyver) [reply properties/test_encode_decode.py]: Aha, I was being distracted by the other errors around the real one. Let's see if the latest commit helps.

**Comment by takluyver** [comment]
OK, looks like the test failure now is real. Let me know if you want me to comment out the relevant line so the tests pass.

**Comment by max-sixty** [comment]
You're right @takluyver 

It looks like hypothesis tests are running in the normal test suites. Anyone know offhand why that is? e.g. https://dev.azure.com/xarray/xarray/_build/results?buildId=1284

(that doesn't solve the test failure, though)

**Comment by max-sixty** [comment]
> OK, looks like the test failure now is real. Let me know if you want me to comment out the relevant line so the tests pass.

If we want to merge a subset of the tests then that's fine. Ofc even better if we can use these tests to find & fix the errors

**Comment by Zac-HD** [comment]
>> OK, looks like the test failure now is real. Let me know if you want me to comment out the relevant line so the tests pass.

> If we want to merge a subset of the tests then that's fine. Ofc even better if we can use these tests to find & fix the errors

In my experience it's better to open an issue, add an xfail decorator to the test, and merge the tests PR.  Otherwise the initial PR can take a very long time and no other property-based tests get added. 

It's also likely that the person who found the bug is not the best person to fix it, and requiring that they do so in order to merge a useful test just disincentives testing!

**Comment by shoyer** [comment]
> In my experience it's better to open an issue, add an xfail decorator to the test, and merge the tests PR. Otherwise the initial PR can take a very long time and no other property-based tests get added.

+1 let's do that!

**Comment by takluyver** [comment]
OK, I've xfailed it.
